doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,203 @@ entirely in Julia. You can read their source and see precisely what they
+between parsing and run-time. Staged functions give the capability to
+generate specialized code depending on the types of their arguments.
+While macros work with expressions at parsing-time and cannot access the
+types of their inputs, a staged function gets expanded at a time when
+the types of the arguments are known, but the function is not yet compiled.
+
+Depending on the types of the arguments, a staged function returns a quoted
+expression which then forms the function body of the specialized function.
+Thus, staged functions provide a flexible framework to move work from
+run-time to compile-time.
+
+When defining staged functions, there are three main differences to
+ordinary functions:
+
+1. You use the keyword ``stagedfunction`` instead of ``function``
+
+2. In the body of the ``stagedfunction`` you only have access to the
+   *types* of the arguments, not their values.
+
+3. Instead of calculating something or performing some action, you return
+   from the staged function a *quoted expression* which, when evaluated,
+   does what you want.
+
+It's easiest to illustrate this with an example. We can declare a staged
+function ``foo`` as
+
+.. doctest::
+
+    julia> stagedfunction foo(x)
+               println(x)
+               :(x*x)
+           end
+    foo (generic function with 1 method)
+
+Note that the body returns a quoted expression, namely ``x*x``.
+
+From the callers perspective, they are very similar to regular functions;
+in fact, you don't have to know if you're calling a ``function`` or a
+``stagedfunction`` - the syntax and result of the call is just the same.
+Let's see how ``foo`` behaves:
+
+.. doctest::
+
+    julia> x = foo(2); # note: not printing the result
+    Int64              # this is the println() statement in the body
+    julia> x           # now we print x
+    4
+
+    julia> y = foo("bar");
+    ASCIIString
+    julia> y
+    "barbar"
+
+So, we see that in the body of the ``stagedfunction``, ``x`` is the
+*type* of the passed argument, and the value returned by the ``stagedfunction``,
+is the result of evaluating the quoted expression we returned from the
+definition, now with the *value* of ``x``.
+
+What happens if we evaluate ``foo`` again with a type that we have already
+used?
+
+.. doctest::
+
+    julia> foo(4)
+    16
+
+Note that there is no printout of ``Int64``. The body of the ``stagedfunction``
+is only executed *once* (not entirely true, see note below) when the method
+for that specific set of argument types is compiled. After that, the
+expression returned from the ``stagedfunction`` on the first invocation
+is re-used as the method body.
+
+The reason for the disclaimer above is that the number of times a staged
+function is staged is really an implementation detail; it *might* be only
+once, but it *might* also be more often. As a consequence, you should 
+*never* write a staged function with side effects - when, and how often,
+the side effects occur is undefined.
```

Comment:
perhaps it's worth adding a parenthetical here that the same is true of macros.

perhaps it's also worth nothing that, like macros, the use of `eval` in a staged-function is generally an indication that they are being used incorrectly.